File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] a) {
        int max=a[0],m=a[0],min=a[0],temp=0;
        for(int i=1;i<a.length;i++)
        {       
            if(a[i]>0)
            {
                max=Math.max(a[i],max*a[i]);
                min=Math.min(a[i],min*a[i]);
            }
            else if(a[i]==0)
            max=min=0;
            else
            {
                temp=max;
                max=Math.max(a[i],min*a[i]);
                min=Math.min(a[i],temp*a[i]);
            }
            m=Math.max(m,max);
        }
        return m;
        }
}

File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        maxnum=nums[0]
        minnum=nums[0]
        m=nums[0]
        temp=0

        for i in range(1,len(nums)):
            if nums[i] >0:
                maxnum=max(nums[i],maxnum*nums[i])
                minnum=min(nums[i],minnum*nums[i])
            elif nums[i]==0:
                maxnum=0
                minnum=0
            else:
                temp=maxnum
                maxnum=max(nums[i],minnum*nums[i])
                minnum=min(nums[i],temp*nums[i])
            m=max(m,maxnum)
        return m
